src/widgets/concept-widget.tsx

```diff
@@ -0,0 +1,144 @@
+import {
+  VDomModel, VDomRenderer
+} from '@jupyterlab/apputils';
+import {INotebookTracker} from '@jupyterlab/notebook';
+
+import * as React from 'react';
+
+import {AllOfUsConfig} from '../config';
+import {Concept} from '../generated';
+import {ConceptsService} from '../services/concepts.service';
+
+import {Observable} from 'rxjs/Rx';
+
+export class ConceptsWidgetModel extends VDomModel {
+
+  public initialized: boolean;
+  public conceptsLoaded: boolean;
+  public conceptsLoading: boolean;
+  public query = '';
+  public concepts: Array<Concept>;
+  public selectedConcept: Concept;
+
+}
+
+export class ConceptsWidgetRenderer extends VDomRenderer<ConceptsWidgetModel> {
+
+
+  constructor(configObservable: Observable<AllOfUsConfig>,
+              private conceptsService: ConceptsService,
+              private notebooks: INotebookTracker) {
+    super();
+
+    this.id = 'allofus-concepts';
+    this.title.label = 'Concepts';
+    this.title.closable = true;
+    configObservable.subscribe((config) => {
+      if (!this.model.initialized) {
+        this.model.initialized = true;
+        this.update();
+      }
+    });
+  }
+
+  protected render(): React.ReactElement<any> {
+    const handleQueryChange = (event) => {
+      this.model.query = event.target.value;
+      this.update();
+    };
+
+    const searchConcepts = () => {
+      this.model.conceptsLoading = true;
+      this.update();
+      this.conceptsService.searchConcepts({query: this.model.query})
+          .then((concepts) => {
+            this.model.conceptsLoading = false;
+            this.model.concepts = concepts;
+            this.model.conceptsLoaded = true;
+            this.update();
+          });
+    };
+    return (
+        <div>
+          <input value={this.model.query} onChange={handleQueryChange}
+                 disabled={!this.model.initialized}/>
+          <br/>
+          <input value='Search' type='button' onClick={searchConcepts}
+                 disabled={!this.model.initialized || this.model.query == null
+                     || this.model.query === ''}/>
+          <br/>
+          {this.loadingImage()}
+          {this.conceptsTable()}
+        </div>
+    );
+  }
+
+  protected loadingImage(): React.ReactElement<any> {
+    if (!this.model.conceptsLoading) {
+      return null;
+    }
+    const img_src = 'https://vignette.wikia.nocookie.net/epicrapbattlesofhistory/'
+      + 'images/c/c2/Peanut-butter-jelly-time.gif/revision/latest?cb=20141129150614';
```

Comment:
I think this is fun, but should we look for something more consistent with AoU styling?

Reply:
Yes... I don't have such an asset right now. I asked on Slack where the spinner for workbench comes from... I don't see it in our codebase?

Anyway, for now I'm going to leave this in there to remind us to replace it with something. :)

(In general, this UI is a prototype showing what we can do with JupyterLab extensions and React... I don't intend to actually put it in front of researchers in this form.)

Reply:
OK, I wound up swapping this out for a React CSS-based spinner from loaders.css.